dask_match/core.py

```diff
@@ -1,628 +1,279 @@
-            if isinstance(left, Expr):
-                left = left[columns]  # TODO: filter just the correct columns
-
-            if isinstance(right, Expr):
-                right = right[columns]
-
-            return cls(left, right)
-
-        # (a + b)[c] -> a[c] + b[c]
-        yield ReplacementRule(
-            Pattern(cls(left, right)[columns]), functools.partial(transform, cls=cls)
-        )
+    def size(self):
+        return new_collection(self.expr.size)
 
+    def __getitem__(self, other):
+        if isinstance(other, FrameBase):
+            return new_collection(self.expr.__getitem__(other.expr))
+        return new_collection(self.expr.__getitem__(other))
 
-class Add(Binop):
-    operation = operator.add
-    _operator_repr = "+"
+    def __repr__(self):
+        return f"<dask_match.core.Series: expr={self.expr}>"
 
-    @classmethod
-    def _replacement_rules(cls):
-        x = Wildcard.dot("x")
-        yield ReplacementRule(
-            Pattern(Add(x, x)),
-            lambda x: Mul(2, x),
-        )
 
-        yield from super()._replacement_rules()
+class Index(Series):
+    """Index-like Expr Collection"""
 
+    def __repr__(self):
+        return f"<dask_match.core.Index: expr={self.expr}>"
 
-class Sub(Binop):
-    operation = operator.sub
-    _operator_repr = "-"
 
+class Scalar(FrameBase):
+    """Scalar Expr Collection"""
 
-class Mul(Binop):
-    operation = operator.mul
-    _operator_repr = "*"
+    def __repr__(self):
+        return f"<dask_match.core.Scalar: expr={self.expr}>"
 
-    @classmethod
-    def _replacement_rules(cls):
-        a, b, c = map(Wildcard.dot, "abc")
-        yield ReplacementRule(
-            Pattern(
-                Mul(a, Mul(b, c)),
-                CustomConstraint(
-                    lambda a, b, c: isinstance(a, numbers.Number)
-                    and isinstance(b, numbers.Number)
-                ),
-            ),
-            lambda a, b, c: Mul(a * b, c),
+    def __dask_postcompute__(self):
+        return first, ()
+
+
+def new_collection(expr):
+    """Create new collection from an expr"""
+
+    meta = expr._meta
+    if is_dataframe_like(meta):
+        return DataFrame(expr)
+    elif is_series_like(meta):
+        return Series(expr)
+    elif is_index_like(meta):
+        return Index(expr)
+    else:
+        return Scalar(expr)
+
+
+def optimize(collection):
+    from dask_match.expr.core import optimize_expr
+
+    return new_collection(optimize_expr(collection.expr))
+
+
+def from_pandas(*args, **kwargs):
+    from dask_match.expr.io.io import FromPandas
+
+    return new_collection(FromPandas(*args, **kwargs))
+
+
+def from_graph(*args, **kwargs):
+    from dask_match.expr.io.io import FromGraph
+
+    return new_collection(FromGraph(*args, **kwargs))
+
+
+def read_csv(*args, **kwargs):
+    from dask_match.expr.io.csv import ReadCSV
+
+    return new_collection(ReadCSV(*args, **kwargs))
+
+
+def read_parquet(
```

Comment:
Maybe the functions above can move into some namespace module, like `__init__.py` without being in core.py